server/src/server.erl

```diff
@@ -1,142 +1,127 @@
+            joining_map(State, Table, Request);
+        {update_character, CharacterMap} ->
+            update(State, Table, CharacterMap);
+        {Kind, Request} ->
+            io:format("[~p] ~p sent an unsuppoted request: ~p\n", [?SERVER, Kind, Request]);
+        {system, From, Request} ->
+            sys:handle_system_msg(Request, From, Parent, server, Options, State)
     end,
-    handle_master(Connection).
+    loop(Parent, Options, State, Table).
+
+% Game State
+login(From, {Username, Password}, State) ->
+    io:format("[~p] User ~p is attemptig to login from ~p\n", [?SERVER, Username, From]),
+    case registry:check_user(#{username => Username, password => Password},
+                             State#state.connection)
+    of
+        {ok, Email} ->
+            io:format("[~p] USER: ~p successfully logged in~n", [?SERVER, Email]),
+            ets:insert(?MODULE, {Email, From}),
+            io:format("[~p] <Email=~p, PID=~p> inserted...~n", [?SERVER, Email, From]),
+            From ! {ok, {State#state.pid, Email}};
+        {error, Message} ->
+            io:format("Unexpected Error: ~p\n", [Message]),
+            From ! {error, Message}
+    end.
+
+list_characters(State, Table, #{email := Email, username := Username} = Request) ->
+    io:format("[~p] Querying ~p's characters...~n", [?SERVER, Username]),
+    {ok, Pid} = safe_lookup(Table, Email),
+    Reply = character:player_characters(Request, State#state.connection),
+    io:format("[~p] Characters: ~p~n", [?SERVER, Reply]),
+    Pid ! Reply.
+
+joining_map(State,
+            Table,
+            #{username := _Username,
+              email := Email,
+              name := Name} =
+                Request) ->
+    {ok, Pid} = safe_lookup(Table, Email),
+    case character:activate(Request, erlang:pid_to_list(Pid), State#state.connection) of
+        ok ->
+            io:format("[~p] Retriving ~p's updated info...", [?SERVER, Name]),
+            Result = character:player_character(Request, State#state.connection),
+            io:format("Got: ~p~n", [Result]),
+            Pid ! Result;
+        {error, Message} ->
+            io:format("Unexpected Error: ~p\n", [Message]),
+            Pid ! {error, "Could not join map"}
+    end.
 
-stop(_) ->
-    ok.
+update(State, Table, CharacterMap) ->
+    {ok, Pid} = safe_lookup(Table, maps:get(email, CharacterMap)),
+    case character:update(CharacterMap, State#state.connection) of
+        ok ->
+            Result =
+                character:retrieve_near_players(CharacterMap,
+                                                erlang:pid_to_list(Pid),
+                                                State#state.connection),
+            Pid ! Result;
+        {error, Message} ->
+            Pid ! {error, Message}
+    end.
 
-main(_) ->
-    start(none, none).
+safe_lookup(Table, Email) ->
+    case ets:lookup(Table, Email) of
+        [{_, Pid}] ->
+            {ok, Pid};
+        _ ->
+            error
+    end.
```

Comment:
My understanding is that we should do a lookup using the primary of the table:
```sql
CREATE TABLE player.record(
       username VARCHAR(32) NOT NULL,
       password TEXT NOT NULL,
       e_mail TEXT NOT NULL CHECK (e_mail ~* '^[A-Za-z0-9.+%-]+@[A-Za-z0-9.-]+[.][A-Za-z]+$'),
       PRIMARY KEY(username, e_mail)
);
```

Reply:
@EduardoLR10 I know but there is now way as far as I can see to properly decode a string back to a `Pid`.

Reply:
Have you tested `erlang:list_to_pid/1`? I would assume that the PSQL would return binary list from a blob field.

Reply:
let me test this out

Reply:
I does not work, we save a string representation in the DB that can't be converted back, even when you put it as a list or binary.

Reply:
> Are you sure? A string in Erlang is a list of characters afaik... When doing a query of this in the DB you would have to transform it, like we do with `state_type`.

```
erlang:list_to_pid("<14786.0.0>").
** exception error: bad argument
     in function  list_to_pid/1
        called as list_to_pid("<14786.0.0>")
        *** argument 1: not a textual representation of a pid
```

Reply:
```suggestion
safe_lookup(Table, Email) ->
    case ets:lookup(Table, Email) of
        [{_, Pid}] ->
            {ok, Pid};
        _ ->
            error
    end.
```

Also, why are you guys saving a value that can anything with the VM restarting in the DB?

Reply:
About the error: try removing `<` and `>` from there.
About the question: I will assume you are talking about the `active` character table. We are doing it because a PID does not change during a playing session, and we need to know who is playing the game at a given moment, because we need to display other players.

Reply:
let me test a different approach